### ISSUE TITLE
Misleading error message when property raises Exception

### ISSUE THREAD

**Issue by dannyadair** [django-haystack/django-haystack#425]
When a property is indexed, and that property raises an Exception upon (re-)indexing, the error message that is shown is "The model X does not have a model_attr Y". That stems from using hasattr() which in turn calls getattr() which will actually "execute" the code, so there would also be a redundant double-calling of that code.

See also somewhat related issue #419
and documentation of hasattr() - http://docs.python.org/library/functions.html#hasattr

Just in case:

> > > class MyClass(object):
> > > ...     @property
> > > ...     def will_pass(self):
> > > ...         x = 1
> > > ...     @property
> > > ...     def will_fail(self):
> > > ...         x = y
> > > ... 
> > > c = MyClass()
> > > hasattr(c, 'will_pass')
> > > True
> > > hasattr(c, 'will_fail')
> > > False

**Comment by zachsnow** [comment]
A partial solution to the "called twice" issue (which just bit us) would be to eschew `hasattr` in favor of `getattr` with an opaque default at <https://github.com/django-haystack/django-haystack/blob/69bb1db279546137b8b03abefecd13fa214d48fe/haystack/fields.py#L88>, roughly to following modulo some renaming:

```
                UNKOWN = {}
                object = getattr(current_object, attr, UNKNOWN)
                if object is UNKNOWN:
                    raise SearchFieldError("The model '%s' does not have a model_attr '%s'." % (repr(current_object), attr))

                elif object is None:
                    ...
```

We would still have the issue that the property raising an `AttributeError` will be treated as the property being missing, but at least we won't be invoking possibly-expensive properties multiple times unnecessarily.

**Comment by acdha** [comment]
@zachsnow That seems reasonable – any chance you could toss a pull request my way?

**Comment by zachsnow** [comment]
@acdha yeah no worries, I have implemented it but need to figure out a good way to reveal the issue in tests and ensure that nothing else breaks.

**Comment by zachsnow** [comment]
@acdha I haven't had a chance to look at testing, but this seems to do the job: https://github.com/zachsnow/django-haystack/tree/dev/address-hasattr-425
